# CCA and Season wise Potential Created Through Dugwells
# Relation between the three datasets
Again there are three datasets with almost similar names matching the above heading.
Let's just dig deep into these for a start.

In [1]:
import pandas as pd
df1 = pd.read_csv('../../Datasets/CCA and Season wise Potential Created (Permanently not in Use) Through Dugwells for 5th Minor Irrigation Census (2013-14).csv')
df2 = pd.read_csv('../../Datasets/CCA and Season wise Potential Created (Temporary not in Use) Through Dugwells for 5th Minor Irrigation Census (2013-14).csv')
df3 = pd.read_csv('../../Datasets/CCA and Season wise Potential Created Through all Dugwells for 5th Minor Irrigation Census (2013-14).csv')

df1 - Permanently not in use  
df2 - Temporarily not in use  
df3 - All  
Cultivable Command Area (CCA):  
The area which can be irrigated from a scheme and is fit for cultivation. 

In [2]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18616 entries, 0 to 18615
Data columns (total 9 columns):
 #   Column                                    Non-Null Count  Dtype  
---  ------                                    --------------  -----  
 0   State                                     18616 non-null  object 
 1   District                                  18616 non-null  object 
 2   Block/Tehsil                              18616 non-null  object 
 3   Village                                   18616 non-null  object 
 4   Culturable Command Area                   18616 non-null  float64
 5   Irrigation potential created - Kharif     18616 non-null  float64
 6   Irrigation potential created - Rabi       18616 non-null  float64
 7   Irrigation potential created - Perennial  18616 non-null  float64
 8   Irrigation potential created - Other      18616 non-null  float64
dtypes: float64(5), object(4)
memory usage: 1.3+ MB


In [3]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34138 entries, 0 to 34137
Data columns (total 9 columns):
 #   Column                                    Non-Null Count  Dtype  
---  ------                                    --------------  -----  
 0   State                                     34138 non-null  object 
 1   District                                  34138 non-null  object 
 2   Block/Tehsil                              34138 non-null  object 
 3   Village                                   34138 non-null  object 
 4   Culturable Command Area                   34138 non-null  float64
 5   Irrigation potential created - Kharif     34138 non-null  float64
 6   Irrigation potential created - Rabi       34138 non-null  float64
 7   Irrigation potential created - Perennial  34138 non-null  float64
 8   Irrigation potential created - Other      34138 non-null  float64
dtypes: float64(5), object(4)
memory usage: 2.3+ MB


In [4]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 226052 entries, 0 to 226051
Data columns (total 9 columns):
 #   Column                                    Non-Null Count   Dtype  
---  ------                                    --------------   -----  
 0   State                                     226052 non-null  object 
 1   District                                  226052 non-null  object 
 2   Block/Tehsil                              226052 non-null  object 
 3   Village                                   226052 non-null  object 
 4   Culturable Command Area                   226052 non-null  float64
 5   Irrigation potential created - Kharif     226052 non-null  float64
 6   Irrigation potential created - Rabi       226052 non-null  float64
 7   Irrigation potential created - Perennial  226052 non-null  float64
 8   Irrigation potential created - Other      226052 non-null  float64
dtypes: float64(5), object(4)
memory usage: 15.5+ MB


Similar datasets differing only on the basis of number of data points.

In [5]:
[df1.columns == df2.columns, df3.columns == df2.columns]

[array([ True,  True,  True,  True,  True,  True,  True,  True,  True]),
 array([ True,  True,  True,  True,  True,  True,  True,  True,  True])]

Columns are the same too.

In [6]:
df1.head()

State   District Block/Tehsil              Village  \
0  ANDAMAN & NICOBARS    NICOBAR  CAR NICOBAR           Big Lapati   
1  ANDAMAN & NICOBARS    NICOBAR  CAR NICOBAR            Tapoiming   
2      ANDHRA PRADESH  Anantapur        Agali                Agali   
3      ANDHRA PRADESH  Anantapur        Agali  Akkagaladevarahalli   
4      ANDHRA PRADESH  Anantapur        Agali  Hulikeradevarahalli   

   Culturable Command Area  Irrigation potential created - Kharif  \
0                      0.1                                    0.1   
1                      0.6                                    0.4   
2                      0.0                                    0.0   
3                      0.0                                    0.0   
4                      0.0                                    0.0   

   Irrigation potential created - Rabi  \
0                                  0.1   
1                                  0.4   
2                                  0.0   
3                                  0.0   
4                                  0.0   

   Irrigation potential created - Perennial  \
0                                       0.0   
1                                       0.0   
2                                       0.0   
3                                       0.0   
4                                       0.0   

   Irrigation potential created - Other  
0                                   0.0  
1                                   0.0  
2                                   0.0  
3                                   0.0  
4                                   0.0

In [7]:
df2.head()

State                District Block/Tehsil            Village  \
0  ANDAMAN & NICOBARS                 NICOBAR  CAR NICOBAR         Big Lapati   
1  ANDAMAN & NICOBARS  NORTH & MIDDLE ANDAMAN       RANGAT             Adojig   
2  ANDAMAN & NICOBARS  NORTH & MIDDLE ANDAMAN       RANGAT            Betapur   
3  ANDAMAN & NICOBARS  NORTH & MIDDLE ANDAMAN       RANGAT  Ramachandra nagar   
4  ANDAMAN & NICOBARS           SOUTH ANDAMAN   FERRARGUNJ         Namunagarh   

   Culturable Command Area  Irrigation potential created - Kharif  \
0                     0.10                                    0.1   
1                     5.40                                    4.4   
2                     1.13                                    0.0   
3                     0.30                                    0.3   
4                     0.62                                    0.5   

   Irrigation potential created - Rabi  \
0                                 0.10   
1                                 0.40   
2                                 0.41   
3                                 0.30   
4                                 0.50   

   Irrigation potential created - Perennial  \
0                                       0.0   
1                                       0.2   
2                                       0.1   
3                                       0.0   
4                                       0.0   

   Irrigation potential created - Other  
0                                   0.0  
1                                   0.0  
2                                   0.0  
3                                   0.0  
4                                   0.0

In [8]:
df3.head()

State District  Block/Tehsil         Village  \
0  ANDAMAN & NICOBARS  NICOBAR  CAMPBELL BAY    Govind Nagar   
1  ANDAMAN & NICOBARS  NICOBAR  CAMPBELL BAY  Joginder Nagar   
2  ANDAMAN & NICOBARS  NICOBAR  CAMPBELL BAY   Sashtri Nagar   
3  ANDAMAN & NICOBARS  NICOBAR   CAR NICOBAR      Big Lapati   
4  ANDAMAN & NICOBARS  NICOBAR   CAR NICOBAR     Chuckchucha   

   Culturable Command Area  Irrigation potential created - Kharif  \
0                     7.00                                   6.00   
1                     8.00                                   8.00   
2                     5.00                                   5.00   
3                     1.26                                   1.01   
4                     2.51                                   1.99   

   Irrigation potential created - Rabi  \
0                                 5.50   
1                                 8.00   
2                                 5.00   
3                                 0.91   
4                                 0.70   

   Irrigation potential created - Perennial  \
0                                       0.0   
1                                       0.0   
2                                       0.0   
3                                       0.0   
4                                       0.0   

   Irrigation potential created - Other  
0                                   0.0  
1                                   0.0  
2                                   0.0  
3                                   0.0  
4                                   0.0

Again, we'll need a unique index for comparision.

In [9]:
import sys
sys.path.append('../Index/')
import Clean

In [10]:
df1 = Clean.clean(df1)
df2 = Clean.clean(df2)
df3 = Clean.clean(df3)
df2.head()

Culturable Command Area  \
State_District_Block/Tehsil_Village                                           
ANDAMAN & NICOBARS_NICOBAR_CAR NICOBAR_Big Lapati                      0.10   
ANDAMAN & NICOBARS_NORTH & MIDDLE ANDAMAN_RANGA...                     5.40   
ANDAMAN & NICOBARS_NORTH & MIDDLE ANDAMAN_RANGA...                     1.13   
ANDAMAN & NICOBARS_NORTH & MIDDLE ANDAMAN_RANGA...                     0.30   
ANDAMAN & NICOBARS_SOUTH ANDAMAN_FERRARGUNJ_Nam...                     0.62   

                                                    Irrigation potential created - Kharif  \
State_District_Block/Tehsil_Village                                                         
ANDAMAN & NICOBARS_NICOBAR_CAR NICOBAR_Big Lapati                                     0.1   
ANDAMAN & NICOBARS_NORTH & MIDDLE ANDAMAN_RANGA...                                    4.4   
ANDAMAN & NICOBARS_NORTH & MIDDLE ANDAMAN_RANGA...                                    0.0   
ANDAMAN & NICOBARS_NORTH & MIDDLE ANDAMAN_RANGA...                                    0.3   
ANDAMAN & NICOBARS_SOUTH ANDAMAN_FERRARGUNJ_Nam...                                    0.5   

                                                    Irrigation potential created - Rabi  \
State_District_Block/Tehsil_Village                                                       
ANDAMAN & NICOBARS_NICOBAR_CAR NICOBAR_Big Lapati                                  0.10   
ANDAMAN & NICOBARS_NORTH & MIDDLE ANDAMAN_RANGA...                                 0.40   
ANDAMAN & NICOBARS_NORTH & MIDDLE ANDAMAN_RANGA...                                 0.41   
ANDAMAN & NICOBARS_NORTH & MIDDLE ANDAMAN_RANGA...                                 0.30   
ANDAMAN & NICOBARS_SOUTH ANDAMAN_FERRARGUNJ_Nam...                                 0.50   

                                                    Irrigation potential created - Perennial  \
State_District_Block/Tehsil_Village                                                            
ANDAMAN & NICOBARS_NICOBAR_CAR NICOBAR_Big Lapati                                        0.0   
ANDAMAN & NICOBARS_NORTH & MIDDLE ANDAMAN_RANGA...                                       0.2   
ANDAMAN & NICOBARS_NORTH & MIDDLE ANDAMAN_RANGA...                                       0.1   
ANDAMAN & NICOBARS_NORTH & MIDDLE ANDAMAN_RANGA...                                       0.0   
ANDAMAN & NICOBARS_SOUTH ANDAMAN_FERRARGUNJ_Nam...                                       0.0   

                                                    Irrigation potential created - Other  
State_District_Block/Tehsil_Village                                                       
ANDAMAN & NICOBARS_NICOBAR_CAR NICOBAR_Big Lapati                                    0.0  
ANDAMAN & NICOBARS_NORTH & MIDDLE ANDAMAN_RANGA...                                   0.0  
ANDAMAN & NICOBARS_NORTH & MIDDLE ANDAMAN_RANGA...                                   0.0  
ANDAMAN & NICOBARS_NORTH & MIDDLE ANDAMAN_RANGA...                                   0.0  
ANDAMAN & NICOBARS_SOUTH ANDAMAN_FERRARGUNJ_Nam...                                   0.0

In [11]:
pd.Series(df1.index).apply(lambda x : x in df3.index).value_counts()

True    18616
Name: State_District_Block/Tehsil_Village, dtype: int64

In [12]:
pd.Series(df2.index).apply(lambda x : x in df3.index).value_counts()

True    34138
Name: State_District_Block/Tehsil_Village, dtype: int64

Everything in the first two datasets are present in the third one as expected.

In [13]:
pd.Series(df2.index).apply(lambda x : x in df1.index).value_counts()

False    26493
True      7645
Name: State_District_Block/Tehsil_Village, dtype: int64

In [14]:
pd.Series(df1.index).apply(lambda x : x in df2.index).value_counts()

False    10971
True      7645
Name: State_District_Block/Tehsil_Village, dtype: int64

The first two datasets have 7645 intersecting data points.

In [15]:
df3.iloc[0]

Culturable Command Area                     7.0
Irrigation potential created - Kharif       6.0
Irrigation potential created - Rabi         5.5
Irrigation potential created - Perennial    0.0
Irrigation potential created - Other        0.0
Name: ANDAMAN & NICOBARS_NICOBAR_CAMPBELL BAY_Govind Nagar, dtype: float64

In [16]:
df3.loc[df3.iloc[0].name]

Culturable Command Area                     7.0
Irrigation potential created - Kharif       6.0
Irrigation potential created - Rabi         5.5
Irrigation potential created - Perennial    0.0
Irrigation potential created - Other        0.0
Name: ANDAMAN & NICOBARS_NICOBAR_CAMPBELL BAY_Govind Nagar, dtype: float64

In [17]:
def check(x, y):
    if x.name in y.index:
        return x == y.loc[x.name]
df1.apply(check, y = df3, axis = 1).sum(axis = 1).value_counts()

2    8101
5    3149
1    3143
3    2286
0    1870
4      67
dtype: int64

In [18]:
df2.apply(check, y = df3, axis = 1).sum(axis = 1).value_counts()

2    13393
5     8952
1     5089
0     3280
3     3120
4      304
dtype: int64

In [19]:
df2.apply(check, y = df1, axis = 1).sum(axis = 1).value_counts()

0.0    26907
2.0     3687
3.0     1815
1.0     1021
5.0      383
4.0      325
dtype: int64

In [20]:
df1.apply(check, y = df2, axis = 1).sum(axis = 1).value_counts()

0.0    11385
2.0     3687
3.0     1815
1.0     1021
5.0      383
4.0      325
dtype: int64

Those don't give away anything.

In [21]:
li = list()
def add(x, y):
    if x.name in y.index:
        x += y.loc[x.name]
        li.append(x.name)
    return x
df4 = df1.apply(add, y = df2, axis = 1)
df4.append(df2.drop(li))
df4 = df4.sort_index()

In [22]:
def sub(x, y):
    if x.name in y.index:
        x -= y.loc[x.name]
    return x
df5 = df3.apply(sub, y = df4, axis = 1)

In [23]:
df5.head()

Culturable Command Area  \
State_District_Block/Tehsil_Village                                           
ANDAMAN & NICOBARS_NICOBAR_CAMPBELL BAY_Govind ...                     7.00   
ANDAMAN & NICOBARS_NICOBAR_CAMPBELL BAY_Joginde...                     8.00   
ANDAMAN & NICOBARS_NICOBAR_CAMPBELL BAY_Sashtri...                     5.00   
ANDAMAN & NICOBARS_NICOBAR_CAR NICOBAR_Big Lapati                      1.06   
ANDAMAN & NICOBARS_NICOBAR_CAR NICOBAR_Chuckchucha                     2.51   

                                                    Irrigation potential created - Kharif  \
State_District_Block/Tehsil_Village                                                         
ANDAMAN & NICOBARS_NICOBAR_CAMPBELL BAY_Govind ...                                   6.00   
ANDAMAN & NICOBARS_NICOBAR_CAMPBELL BAY_Joginde...                                   8.00   
ANDAMAN & NICOBARS_NICOBAR_CAMPBELL BAY_Sashtri...                                   5.00   
ANDAMAN & NICOBARS_NICOBAR_CAR NICOBAR_Big Lapati                                    0.81   
ANDAMAN & NICOBARS_NICOBAR_CAR NICOBAR_Chuckchucha                                   1.99   

                                                    Irrigation potential created - Rabi  \
State_District_Block/Tehsil_Village                                                       
ANDAMAN & NICOBARS_NICOBAR_CAMPBELL BAY_Govind ...                                 5.50   
ANDAMAN & NICOBARS_NICOBAR_CAMPBELL BAY_Joginde...                                 8.00   
ANDAMAN & NICOBARS_NICOBAR_CAMPBELL BAY_Sashtri...                                 5.00   
ANDAMAN & NICOBARS_NICOBAR_CAR NICOBAR_Big Lapati                                  0.71   
ANDAMAN & NICOBARS_NICOBAR_CAR NICOBAR_Chuckchucha                                 0.70   

                                                    Irrigation potential created - Perennial  \
State_District_Block/Tehsil_Village                                                            
ANDAMAN & NICOBARS_NICOBAR_CAMPBELL BAY_Govind ...                                       0.0   
ANDAMAN & NICOBARS_NICOBAR_CAMPBELL BAY_Joginde...                                       0.0   
ANDAMAN & NICOBARS_NICOBAR_CAMPBELL BAY_Sashtri...                                       0.0   
ANDAMAN & NICOBARS_NICOBAR_CAR NICOBAR_Big Lapati                                        0.0   
ANDAMAN & NICOBARS_NICOBAR_CAR NICOBAR_Chuckchucha                                       0.0   

                                                    Irrigation potential created - Other  
State_District_Block/Tehsil_Village                                                       
ANDAMAN & NICOBARS_NICOBAR_CAMPBELL BAY_Govind ...                                   0.0  
ANDAMAN & NICOBARS_NICOBAR_CAMPBELL BAY_Joginde...                                   0.0  
ANDAMAN & NICOBARS_NICOBAR_CAMPBELL BAY_Sashtri...                                   0.0  
ANDAMAN & NICOBARS_NICOBAR_CAR NICOBAR_Big Lapati                                    0.0  
ANDAMAN & NICOBARS_NICOBAR_CAR NICOBAR_Chuckchucha                                   0.0

In [24]:
df3.applymap(lambda x : '-ve' if x<0 else '0' if x==0 else '+ve').stack().value_counts()

+ve    690467
0      439793
dtype: int64

In [25]:
df5.applymap(lambda x : '-ve' if x<0 else '0' if x==0 else '+ve').stack().value_counts()

+ve    680905
0      449090
-ve       265
dtype: int64